File: src/api/message.ts
import axios from "axios";
import { API_URL } from "../config";

// 消息接口
export interface Message {
  id: number;
  sender_id: number;
  receiver_id: number;
  content: string;
  is_read: boolean;
  created_at: string;
  sender: {
    id: number;
    username: string;
    avatar: string;
  };
}

// 获取请求头
const getHeaders = () => {
  const token = localStorage.getItem("token");
  return {
    Authorization: `Bearer ${token}`,
  };
};

// 创建消息
export const createMessage = async (receiverId: number, content: string) => {
  try {
    const response = await axios.post(
      `${API_URL}/messages`,
      {
        receiver_id: receiverId,
        content,
      },
      {
        headers: getHeaders(),
      }
    );
    return response.data;
  } catch (error) {
    console.error("创建消息失败:", error);
    throw error;
  }
};

// 获取消息列表
export const getMessages = async () => {
  try {
    const response = await axios.get(`${API_URL}/messages`, {
      headers: getHeaders(),
    });
    return response.data;
  } catch (error) {
    console.error("获取消息列表失败:", error);
    throw error;
  }
};

// 获取与特定用户的对话
export const getConversation = async (userId: number) => {
  try {
    const response = await axios.get(`${API_URL}/messages/${userId}`, {
      headers: getHeaders(),
    });
    return response.data;
  } catch (error) {
    console.error("获取对话失败:", error);
    throw error;
  }
};

// 标记消息已读
export const markMessageAsRead = async (messageId: number) => {
  try {
    const response = await axios.put(
      `${API_URL}/messages/${messageId}/read`,
      {},
      {
        headers: getHeaders(),
      }
    );
    return response.data;
  } catch (error) {
    console.error("标记消息已读失败:", error);
    throw error;
  }
};

// 获取未读消息数量
export const getUnreadCount = async () => {
  try {
    const response = await axios.get(`${API_URL}/messages/unread/count`, {
      headers: getHeaders(),
    });
    return response.data;
  } catch (error) {
    console.error("获取未读消息数量失败:", error);
    throw error;
  }
};

File: backend/app/routes/message.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from app.core.auth import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.models.message import Message
from app.schemas.message import MessageCreate, MessageResponse
from datetime import datetime

router = APIRouter()

@router.post("/messages", response_model=MessageResponse)
async def create_message(
    message: MessageCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """创建新消息"""
    db_message = Message(
        sender_id=current_user.id,
        receiver_id=message.receiver_id,
        content=message.content
    )
    db.add(db_message)
    db.commit()
    db.refresh(db_message)
    return db_message

@router.get("/messages", response_model=List[MessageResponse])
async def get_messages(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取当前用户的所有消息"""
    messages = db.query(Message).filter(
        (Message.sender_id == current_user.id) | 
        (Message.receiver_id == current_user.id)
    ).order_by(Message.created_at.desc()).all()
    return messages

@router.get("/messages/{user_id}", response_model=List[MessageResponse])
async def get_conversation(
    user_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取与特定用户的对话"""
    messages = db.query(Message).filter(
        ((Message.sender_id == current_user.id) & (Message.receiver_id == user_id)) |
        ((Message.sender_id == user_id) & (Message.receiver_id == current_user.id))
    ).order_by(Message.created_at.asc()).all()
    return messages

@router.put("/messages/{message_id}/read")
async def mark_message_as_read(
    message_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """将消息标记为已读"""
    message = db.query(Message).filter(
        Message.id == message_id,
        Message.receiver_id == current_user.id
    ).first()
    
    if not message:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Message not found"
        )
    
    message.is_read = True
    db.commit()
    return {"status": "success"}

@router.get("/messages/unread/count")
async def get_unread_count(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取未读消息数量"""
    count = db.query(Message).filter(
        Message.receiver_id == current_user.id,
        Message.is_read == False
    ).count()
    return {"unread_count": count} 